hub/api/dataset.py

```diff
@@ -68,6 +69,14 @@ def __init__(
         else:
             self.meta = {"tensors": []}
 
+    def __enter__(self):
```

Comment:
does this mean:

```python
with Dataset(...) as ds:
    # __enter__ is called before entry
```

?

Reply:
ah i see it is in the tests.

maybe this isn't the best way, it's unclear.

```python
ds = Dataset(...)
tensor = ds.tensor

tensor.append(...)  # auto flush

with ds.disable_autoflush():
    tensor.append(...)  # no flush
    tensor.append(...)  # no flush
    tensor.append(...)  # no flush

# ds.flush()  # <-- not necessary, called after `with ds.disable_autoflush()` __exit__ 
```

Reply:
@AbhinavTuli @benchislett  thoughts?

Reply:
https://pytorch.org/docs/stable/generated/torch.no_grad.html

it may be worth it to mimic this syntax:

```python
ds1 = Dataset(...)
ds2 = Dataset(...)
tensor1 = ds1.tensor
tensor2 = ds2.tensor

tensor1.append(...)  # auto flush
tensor2.append(...)  # auto flush

with hub.disable_autoflush():
    tensor1.append(...)  # no flush
    tensor2.append(...)  # no flush
    tensor1.append(...)  # no flush

# ds1.flush()  # <-- not necessary, called after `with ds.disable_autoflush()` __exit__ 
# ds2.flush()  # <-- not necessary, called after `with ds.disable_autoflush()` __exit__ 
```

this is kind of hard because `hub.disable_autoflush()` doesn't take a dataset object, but i think we should have this work.

2 datasets being used at the same time may seem like an edge case, but it's how pytorch handles it and this would be more familiar

Reply:
@McCrearyD 
https://stackoverflow.com/questions/9282967/how-to-open-a-file-using-the-open-with-statement
With supports multiple arguments, apparently.

I still am strongly in favor of `with Dataset(...) as ds` mirroring python's file-system style `with open(...) as file`
They feel like _exactly_ the same use-case, I don't see why we would want to do things differently.